src/lib/y2partitioner/format_mount/base.rb

```diff
@@ -63,19 +65,123 @@ def apply_format_options!
       end
 
       def apply_mount_options!
-        return false unless @partition.filesystem
-
-        if @options.mount
-          @partition.filesystem.mount_point = @options.mount_point
-          @partition.filesystem.mount_by = @options.mount_by
-          @partition.filesystem.label = @options.label if @options.label
-          @partition.filesystem.fstab_options = @options.fstab_options
-        else
-          @partition.filesystem.mount_point = ""
-        end
+        return false if filesystem.nil?
+
+        set_mount_point
+        set_mount_options
 
         true
       end
+
+    private
+
+      attr_reader :partition
+
+      attr_reader :options
+
+      def set_mount_point
+        return unless change_mount_point?
+
+        subvolume_actions_before_set_mount_point
+        filesystem.mount_point = mount? ? options.mount_point : ""
+        subvolume_actions_after_set_mount_point
+      end
+
+      def set_mount_options
+        return unless mount?
+
+        filesystem.mount_by = options.mount_by
+        filesystem.label = options.label if options.label
+        filesystem.fstab_options = options.fstab_options
+      end
+
+      def subvolume_actions_before_set_mount_point
+        delete_new_subvolumes if btrfs?
+      end
+
+      def subvolume_actions_after_set_mount_point
+        add_proposed_subvolumes if btrfs_root?
+        update_mount_points if btrfs?
+        Y2Storage::Filesystems::Btrfs.refresh_root_subvolumes_shadowing(devicegraph)
+      end
+
+      def delete_new_subvolumes
+        new_subvolumes.each { |s| filesystem.delete_btrfs_subvolume(devicegraph, s.path) }
+      end
+
+      def new_subvolumes
+        devicegraph = DeviceGraphs.instance.system
+        filesystem.btrfs_subvolumes.select { |s| !s.exists_in_devicegraph?(devicegraph) }
+      end
+
+      def add_proposed_subvolumes
+        specs = Y2Storage::SubvolSpecification.for_current_product
+
+        specs.each do |spec|
+          next if exist_subvolume?(spec)
+          add_proposed_subvolume(spec)
+        end
+      end
+
+      def exist_subvolume?(spec)
+        path = filesystem.btrfs_subvolume_path(spec.path)
+        !filesystem.find_btrfs_subvolume_by_path(path).nil?
+      end
+
+      def add_proposed_subvolume(spec)
```

Comment:
My gut feelings tell me that this and `Planned::BtrfsSubvolume#create_subvol` should share something else than 4 lines of code.